src/runtime_src/core/tools/common/ReportMemory.cpp

```diff
@@ -0,0 +1,447 @@
+// Local - Include Files
+#include "ReportMemory.h"
+#include <boost/lexical_cast.hpp>
+#include <boost/algorithm/string.hpp>
+#include "core/common/query_requests.h"
+#include "core/common/device.h"
+#include "core/common/utils.h"
+namespace qr = xrt_core::query;
+
+#define XCL_NO_SENSOR_DEV_LL    ~(0ULL)
+#define XCL_NO_SENSOR_DEV       ~(0U)
+#define XCL_NO_SENSOR_DEV_S     0xffff
+#define XCL_INVALID_SENSOR_VAL 0
+
+static const std::map<MEM_TYPE, std::string> memtype_map = {
+    {MEM_DDR3, "MEM_DDR3"},
+    {MEM_DDR4, "MEM_DDR4"},
+    {MEM_DRAM, "MEM_DRAM"},
+    {MEM_STREAMING, "MEM_STREAMING"},
+    {MEM_PREALLOCATED_GLOB, "MEM_PREALLOCATED_GLOB"},
+    {MEM_ARE, "MEM_ARE"},
+    {MEM_HBM, "MEM_HBM"},
+    {MEM_BRAM, "MEM_BRAM"},
+    {MEM_URAM, "MEM_URAM"},
+    {MEM_STREAMING_CONNECTION, "MEM_STREAMING_CONNECTION"}
+};
+
+template <typename T>
+inline std::string pretty( const T &val, const std::string &default_val = "N/A", bool isHex = false )
+{   
+    if (typeid(val).name() != typeid(std::string).name()) {
+        if (val >= std::numeric_limits<T>::max() || val == 0) {
+            return default_val;
+        }
+
+        if (isHex) {
+            std::stringstream ss;
+            ss << "0x" << std::hex << val;
+            return ss.str();
+        }
+    }
+    return boost::lexical_cast<std::string>(val);
+}
+
+static int eccStatus2Str(uint64_t status, std::string& str) 
+{    
+  const int ce_mask = (0x1 << 1);
+  const int ue_mask = (0x1 << 0);
+
+  str.clear();
+
+  // If unknown status bits, can't support.
+  if (status & ~(ce_mask | ue_mask)) {
+    //Bad ECC status detected!
+    return -EINVAL;
+  }    
+
+  if (status == 0) { 
+    str = "(None)";
+    return -EINVAL;
+  }    
+
+  if (status & ue_mask)
+    str += "UE ";
+  if (status & ce_mask)
+    str += "CE ";
+  // Remove the trailing space.
+  str.pop_back();
+  return 0;
+}
+
+void getChannelinfo(const xrt_core::device * device, boost::property_tree::ptree& pt) {
+  std::vector<std::string> dma_threads;
+  boost::property_tree::ptree pt_dma_array;
+  try {
+    dma_threads = xrt_core::device_query<qr::dma_threads_raw>(device);
+  } catch (const std::exception& ex){
+    pt.put("error_msg", ex.what());
+  }
+  uint64_t h2c[8];
+  uint64_t c2h[8];
+  for (unsigned i = 0; i < dma_threads.size(); i++) {
+    boost::property_tree::ptree pt_dma;
+    std::stringstream ss(dma_threads[i]);
+    ss >> c2h[i] >> h2c[i];
+    pt_dma.put("channel_id", i);
+    pt_dma.put("host_to_card_bytes", boost::format("0x%x") % h2c[i]);
+    pt_dma.put("card_to_host_bytes", boost::format("0x%x") % c2h[i]);
+    pt_dma_array.push_back(std::make_pair("",pt_dma));
+  }
+  pt.put(std::string("board.direct_memory_accesses.type"), "pcie xdma");
+  pt.add_child(std::string("board.direct_memory_accesses.metrics"), pt_dma_array);
+}
+
+boost::property_tree::ptree
+populate_memtopology(const xrt_core::device * device, const std::string& desc)

```

Comment:
This function is much too long, please refactor.  Long functions are incomprehensible, small functions are self documenting.
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-single